### PR TITLE
[WIP] Swagger Docs, Token and URL Resource Unit Tests

### PR DESCRIPTION
**Pull request by nurav**

### CODE REVIEW

File: ursh/blueprints/api/resources.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def post(self, **kwargs):
         if Token.query.filter_by(name=kwargs['name']).count() != 0:
             raise Conflict({'message': 'Token with name exists', 'args': ['name']})
         new_token = Token()
-        if not validate_callback_url(kwargs.get('callback_url')):
+        if not None and not validate_url(kwargs.get('callback_url')):
```
Comment (ThiefMaster) [inline ursh/blueprints/api/resources.py]: `if not None` makes no sense. That's always `True`...

File: ursh/blueprints/api/resources.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ def patch(self, api_key, **kwargs):
         token = Token.query.filter_by(api_key=api_key).one_or_none()
         if token is None:
             raise NotFound({'message': 'API key does not exist', 'args': ['api_key']})
-        if not validate_callback_url(kwargs.get('callback_url')):
+        if not None and not validate_url(kwargs.get('callback_url')):
```
Comment (ThiefMaster) [inline ursh/blueprints/api/resources.py]: same

File: ursh/blueprints/api/resources.py [hunk 3]
```diff
@@ -84,7 +84,8 @@ class URLResource(MethodResource):
     @marshal_with(URLSchema(strict=True), code=201)
     @use_kwargs(URLSchema)
     def post(self, **kwargs):
-        if kwargs.get('url') is None:
+        posted_url = kwargs.get('url')
+        if posted_url is None or not validate_url(posted_url):
```
Comment (ThiefMaster) [inline ursh/blueprints/api/resources.py]: not directly related to this PR, but i think you can do this on the schema level (using a `validate` callback on the field in the schema)

File: ursh/blueprints/api/resources.py [hunk 4]
```diff
@@ -124,35 +125,47 @@ def patch(self, shortcut, **kwargs):
...
+                filters.append(URL.token == g.token)
             if metadata:
                 filters.append(URL.custom_data.contains(metadata))
             return URL.query.filter(*filters).all()
         else:
-            return URL.query.filter_by(shortcut=shortcut).first_or_404()
+            url = URL.query.filter_by(shortcut=shortcut).one_or_none()
+            if not url:
+                raise NotFound({'message': 'Shortcut does not exist', 'args': ['shortcut']})
+            else:
```
Comment (ThiefMaster) [inline ursh/blueprints/api/resources.py]: not needed, and you can outdent the next line -- if an exception is raised, the code after that won't run ;)

File: ursh/testing/conftest.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+@pytest.fixture(scope='session')
+def postgresql():
+    """Provides a clean temporary PostgreSQL server/database.
+    If the environment variable `URLSHORTENER_TEST_DATABASE_URI` is set, this fixture
+    will do nothing and simply return the connection string from that variable
+    """
+
+    # Use existing database
+    if 'URLSHORTENER_TEST_DATABASE_URI' in os.environ:
```
Comment (ThiefMaster) [inline ursh/testing/conftest.py]: `URSH_TEST_DATABASE_URI`

File: ursh/testing/conftest.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+@pytest.fixture(scope='session')
+def database(app, postgresql):
+    """Creates a test database which is destroyed afterwards
+    Used only internally, if you need to access the database use `db` instead to ensure
+    your modifications are not persistent!
+    """
+    app.config['SQLALCHEMY_DATABASE_URI'] = postgresql
+    db_.init_app(app)
+    if 'URLSHORTENERTEST_DATABASE_URI' in os.environ and os.environ.get('URLSHORTENER_TEST_DATABASE_HAS_TABLES') == '1':
```
Comment (ThiefMaster) [inline ursh/testing/conftest.py]: typo, also `URSH_TEST_DATABASE_URI`

Comment (ThiefMaster) [inline ursh/testing/conftest.py]: i think we can get rid of the DATABASE_HAS_TABLES part. even in indico it's not really used

File: ursh/testing/conftest.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    monkeypatch.setattr(database.session, 'commit', database.session.flush)
+    monkeypatch.setattr(database.session, 'remove', lambda: None)
+    yield database
+    database.session.rollback()
+    database.session.remove()
+
+
+@pytest.fixture
+@pytest.mark.usefixtures('db')
+def count_queries():
```
Comment (ThiefMaster) [inline ursh/testing/conftest.py]: not used -> no need to keep it here. you can always copy it from indico if you end up needing it later (but i kind of doubt it)

File: ursh/blueprints/api/resources.py [hunk 8]
```diff
@@ -1,7 +1,7 @@
 import re
 from uuid import UUID
 
-from flask import Response, g
+from flask import Response, g, jsonify
```
Comment (pferreir) [inline ursh/blueprints/api/resources.py]: `jsonify` not actually used

File: ursh/blueprints/api/handlers.py [hunk 9]
```diff
@@ -2,6 +2,7 @@
 
 
 def handle_bad_requests(error):
+    print(error)
```
Comment (pferreir) [inline ursh/blueprints/api/handlers.py]: I guess this is not intended?

File: ursh/testing/conftest.py (hunk 5 above)
Comment (pferreir) [inline ursh/testing/conftest.py]: I think you forgot this one!

**Comment by nurav** [comment]
@pferreir @ThiefMaster is this ready to be merged?

**Comment by ThiefMaster** [comment]
go for it

**Pull request merged**